kep/9-capacity-scheduling-for-ml-dl-workloads-based-on-scheduler-framework/README.md

```diff
@@ -0,0 +1,208 @@
+
+1. The slack between "min" and "max" can help to tolerate runtime failures. For example, if a Pod that consumes the entire "min" fails to run, new Pods can still be created if the "max" has not been reached. When using Kubernetes resource quota, once the Pod that consumes the entire quota is created, no other Pods can be created even if the Pod fails to run (e.g., the Pod stucks at the image pulling phase). 
+
+2. Improve overall resource utilization by allowing one queue user to "borrow" unused reserved "min" resources from other queues. A queue user's unused "min" resource can be used by other users, under the condition that there is a mechanism to guarantee the "victim" user can consume his "min" resource whenever he needs. Typically, this is done by implementing preemption.
+
+The success of “Queue” based capacity scheduling relies on a proper Pod preemption algorithm implementation. This KEP proposes the minimal scheduler extension to support the “Queue” based scheduling based on the scheduler framework.
+
+
+### Goals
+
+Use a scheduler plugin to support a quota based preemption algorithm. 
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP?  Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+A  "CRD"is needed to interact with the scheduler. The CRD kind name is subject to change. We use "Queue" in the proposal. The CRD defines the min and max allocation for the managed resources such as cpu, memory or extended resources. "Queue" is namespace scoped. This version of proposal does not consider merging the queue into the existing ResourceQuota. A sample yaml is listed below:
+
+```yaml
+apiVersion: v1
+kind: Queue
+metadata:
+  name: test
+  namespace: test
+spec:
+  max:
+    cpu: 20
+    memory: 40Gi
+    nvidia.com/gpu: 2
+  min:
+    cpu: 10
+    memory: 20Gi
+    nvidia.com/gpu: 1
+```
+
+The definitions of "min" and "max" have been explained above. Max is infinite by default, min is 0 by default. There is an admission webhook to ensure the aggregate resource configurations of all the Pods that consume a Queue does not exceed the Queue’s max resource. Again, the sum of min in all Queue must be less than or equal to the total resource capacity of the cluster. This is enforced by a separate admission webhook. The Queue controller will track the resource allocation.
+
+The Queue objects are created by a cluster administrator. One Queue object per namespace (this assumption can be relaxed if we later introduce a “label” to match Queue CR and Queue consumers).
+
```

Comment:
I think we miss a section here to explain the following scenarios:

- when and how resources are borrowed
- when and how resources are returned back

Reply:
And I would suggest we use a Pod's lifecycle to connect the dots into a whole story.

Reply:
I  add more detail in `PostFilter` to introduce `borrow` and `return`.